keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,896 @@
+respectively to the type
+
+If the client did not specify the preferred notification granularity, default values can be used.
+
+This will be easy to use in CRI-compliant command-line-tools as well as kubelet to monitor the
+progress and publish events to the Pod object
+
+See design detail below.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The user is deploying an application to the k8s cluster, the Pod is being
+scheduled and creation of the Pod beginning. The only event occuring on the Pod
+object events list is "Pulling image", with no progress indicator, or ETA.
+
+If CRI had a possility to expose the progress and / or ETA for completing the
+image pull, it would improve user experience greatly as well as debugging
+experience in situations when the image source is not available and the timeout
+will be inevitably reached, but not immediately: the progress events on the Pod
+object would indicate if the image transfer is ongoing at all.
+
+#### Story 2
+
+The user is debugging an issue in k8s cluster by running `crictl pull` command
+to pull the image - either to see if image is available or to check if connectivity
+to the image source is in place. If the image is available and the pulling
+is ongoing - no indication is possible to show based on CRI itself. At least
+containerd runtime is already providing such information based on the
+`ctr image pull docker.io/library/busybox:latest`.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The risks are minimal, introduction of new API is not affecting any other APIs. Implementation of
+kubelet taking in use the new API to monitor and publish image pulling progress events to Pod
+object are out of scope of this KEP, although kept in mind during design.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Following new CRI streaming API call is proposed:
```

Comment:
I think it might be better to switch to polling, using streaming can cause congestion in case of a large number of started containers, and we actually don't pay much attention to pull progress in most cases.

In crictl, I recommend polling for pulling progress

In kubectl, I suggest setting up a periodic poll to check the progress of the images that the current node needs but has not yet succeeded in sending as an event.

Just my side opinion

Reply:
Good point, thanks.